### PR TITLE
Bump Rocketchip to June 2020 for Tile changes

### PR DESCRIPTION
**Pull request by jerryz123**
Rocketchip bump to grab generic Tile instantiation changes.

**Related issue**: 
https://github.com/chipsalliance/rocket-chip/pull/2504

**Type of change**: new feature 

**Impact**: rtl change

**Release Notes**

- Legacy references to `XTilesKey` are removed everywhere
- TraceGen, BOOM, and Ariane tiles now use `TilesLocated` key 
- Chipyard `System`/`Subsystem` renamed to `ChipyardSystem`/`ChipyardSubsystem` to avoid name conflict with RC
- `WithControlCore` removed (this was duplicate of `WithNSmallCores`).
- `ChipyardSubsystem` now uses generic `HasTiles` to instantiate tiles

Submodule bumps: 

- [x] https://github.com/firesim/firesim/pull/586
- [x] https://github.com/ucb-bar/ariane-wrapper/pull/7
- [x] https://github.com/ucb-bar/testchipip/pull/89
- [x] https://github.com/ucb-bar/dsptools/pull/208
- [x] https://github.com/riscv-boom/riscv-boom/pull/470

### CODE REVIEW

File: generators/chipyard/src/main/scala/ConfigFragments.scala [hunk 1]
```diff
@@ -117,38 +109,23 @@ class WithMultiRoCCHwacha(harts: Int*) extends Config((site, here, up) => {
 })
 
 
-/**
- * Config fragment to add a small Rocket core to the system as a "control" core.
- * Used as an example of a PMU core.
- */
-class WithControlCore extends Config((site, here, up) => {
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/ConfigFragments.scala:124]: Is this not possible to represent in the new system?

File: generators/chipyard/src/main/scala/ConfigFragments.scala [hunk 2]
```diff
@@ -117,38 +109,23 @@ class WithMultiRoCCHwacha(harts: Int*) extends Config((site, here, up) => {
...
+    case tp: ArianeTileAttachParams => tp.copy(tileParams = tp.tileParams.copy(
+      trace = true))
+    case other => other
+  }
   case TracePortKey => Some(TracePortParams())
 })
+
+class WithNPerfCounters(n: Int = 29) extends Config((site, here, up) => {
+  case TilesLocated(InSubsystem) => up(TilesLocated(InSubsystem), site) map {
+    case tp: BoomTileAttachParams => tp.copy(tileParams = tp.tileParams.copy(
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/ConfigFragments.scala:125]: Keep rocket first.
Also we don't need to replicate 29 so many times just use `n` which has the default below.

File: generators/chipyard/src/main/scala/ConfigFragments.scala [hunk 3]
```diff
@@ -117,38 +109,23 @@ class WithMultiRoCCHwacha(harts: Int*) extends Config((site, here, up) => {
...
-      hartId = up(RocketTilesKey, site).size + up(BoomTilesKey, site).size
-    )
-  case MaxHartIdBits => log2Up(up(RocketTilesKey, site).size + up(BoomTilesKey, site).size + 1)
-})
-
 class WithTraceIO extends Config((site, here, up) => {
-  case BoomTilesKey => up(BoomTilesKey) map (tile => tile.copy(trace = true))
-  case ArianeTilesKey => up(ArianeTilesKey) map (tile => tile.copy(trace = true))
+  case TilesLocated(InSubsystem) => up(TilesLocated(InSubsystem), site) map {
+    case tp: BoomTileAttachParams => tp.copy(tileParams = tp.tileParams.copy(
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/ConfigFragments.scala:114]: This is different than rocket's `traceNode`?

File: generators/chipyard/src/main/scala/Subsystem.scala [hunk 4]
```diff
@@ -21,93 +21,36 @@ import freechips.rocketchip.util._
 import freechips.rocketchip.subsystem._
 import freechips.rocketchip.amba.axi4._
 
-import boom.common.{BoomTile, BoomTilesKey, BoomCrossingKey, BoomTileParams}
-import ariane.{ArianeTile, ArianeTilesKey, ArianeCrossingKey, ArianeTileParams}
+import boom.common.{BoomTile}
 
-import testchipip.{DromajoHelper}
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/Subsystem.scala:27]: 🎉

File: generators/chipyard/src/main/scala/TestSuites.scala [hunk 5]
```diff
@@ -109,37 +113,41 @@ class TestSuiteHelper
...
-          addSuites(env.map(rv32uf))
-          if (cfg.fLen >= 64)
-            addSuites(env.map(rv32ud))
-        } else {
-          addSuite(rv32udBenchmarks)
-          addSuites(env.map(rv64uf))
-          if (cfg.fLen >= 64)
-            addSuites(env.map(rv64ud))
+    p(TilesLocated(InSubsystem)).find(_.tileParams.hartId == 0).map {
+      case tp: RocketTileAttachParams => {
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/TestSuites.scala:117]: These seem quite redundant to me but that refactor is definitely for another PR.

File: generators/chipyard/src/main/scala/config/BoomConfigs.scala [hunk 6]
```diff
@@ -96,29 +92,7 @@ class DualSmallBoomConfig extends Config(
   new freechips.rocketchip.subsystem.WithNoSlavePort ++
   new freechips.rocketchip.subsystem.WithInclusiveCache ++
   new freechips.rocketchip.subsystem.WithNExtTopInterrupts(0) ++
-  new boom.common.WithSmallBooms ++
-  new boom.common.WithNBoomCores(2) ++                         // 2 boom cores
-  new freechips.rocketchip.subsystem.WithCoherentBusTopology ++
-  new freechips.rocketchip.system.BaseConfig)
-
-class SmallRV32BoomConfig extends Config(
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/config/BoomConfigs.scala:104]: Why delete this config?

File: generators/chipyard/src/main/scala/config/HeteroConfigs.scala [hunk 7]
```diff
@@ -50,28 +46,6 @@ class HwachaLargeBoomAndHwachaRocketConfig extends Config(
   new freechips.rocketchip.subsystem.WithCoherentBusTopology ++
   new freechips.rocketchip.system.BaseConfig)
 // DOC include end: BoomAndRocketWithHwacha
-
-class DualLargeBoomAndRocketConfig extends Config(
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/config/HeteroConfigs.scala:54]: Why delete this config?

File: generators/chipyard/src/main/scala/config/HeteroConfigs.scala [hunk 8]
```diff
@@ -99,30 +71,6 @@ class LargeBoomAndHwachaRocketConfig extends Config(
 // DOC include end: DualBoomAndRocketOneHwacha
 
 
-
-class LargeBoomAndRV32RocketConfig extends Config(
```
Comment (colinschmidt) [inline generators/chipyard/src/main/scala/config/HeteroConfigs.scala:103]: Why delete this config?

File: generators/tracegen/src/main/scala/Configs.scala [hunk 9]
```diff
@@ -3,87 +3,126 @@ package tracegen
...
+    val prev = up(TilesLocated(InSubsystem), site)
+    val idOffset = overrideIdOffset.getOrElse(prev.size)
+    val memOffset: BigInt = overrideMemOffset.orElse(site(ExtMem).map(_.master.base)).getOrElse(0x0L)
+    params.zipWithIndex.map { case (dcp, i) =>
+      TraceGenTileAttachParams(
+        tileParams = TraceGenParams(
+          hartId = i + idOffset,
+          dcache = Some(dcp),
+          wordBits = site(XLen),
+          addrBits = 32,
```
Comment (colinschmidt) [inline generators/tracegen/src/main/scala/Configs.scala:30]: Why decrease this to 32?

File: generators/chipyard/src/main/scala/ConfigFragments.scala (hunk 1 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/ConfigFragments.scala:124]: This  seems to be equivalent to `WithNSmallCores(1)` in the new system

File: generators/chipyard/src/main/scala/ConfigFragments.scala (hunk 3 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/ConfigFragments.scala:114]: Boom prefers parameterization over whether or not to enable tracing, as it adds significant overhead, while rocket drives its `traceNode` unconditionally.

File: generators/chipyard/src/main/scala/config/BoomConfigs.scala (hunk 6 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/BoomConfigs.scala:104]: Supporting Rv32 in boom is too annoying.

File: generators/chipyard/src/main/scala/config/HeteroConfigs.scala (hunk 7 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/HeteroConfigs.scala:54]: The new `WithNCore` fragments compose more intuitively, such that it feels redundant to have all of these different configs. Heterogeneity is intuitive now.

File: generators/tracegen/src/main/scala/Configs.scala (hunk 9 above)
Comment (jerryz123) [reply generators/tracegen/src/main/scala/Configs.scala:30]: My mistake, I copied this over from RocketChip's tracegen Config, which sets it to 32.

File: generators/chipyard/src/main/scala/TestSuites.scala (hunk 5 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/TestSuites.scala:117]: I agree with both statements

File: generators/chipyard/src/main/scala/ConfigFragments.scala (hunk 1 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/ConfigFragments.scala:124]: I fixed the ControlCore config to place the SmallRocket in the last hartid

File: generators/chipyard/src/main/scala/config/HeteroConfigs.scala (hunk 8 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/HeteroConfigs.scala:103]: Also I don't think this config does what it describes. I'm pretty sure it instantiates 2x RV64 booms and 1x RV64 rocket

File: generators/chipyard/src/main/scala/ConfigFragments.scala (hunk 1 above)
Comment (colinschmidt) [reply generators/chipyard/src/main/scala/ConfigFragments.scala:124]: Great thanks.

File: generators/chipyard/src/main/scala/config/HeteroConfigs.scala (hunk 7 above)
Comment (colinschmidt) [reply generators/chipyard/src/main/scala/config/HeteroConfigs.scala:54]: 👍

**Comment by davidbiancolin** [comment]
Are you planning to test this in FireSim?

**Comment by jerryz123** [comment]
> Are you planning to test this in FireSim?

I can push a rocket config through, but I wasn't planning on regenerating AGFIs, since no FireSim changes were necessary.

**Comment by davidbiancolin** [comment]
Considering there are 234 new commits i want to see an AGFI regeneration + linux boots.

**Comment by jerryz123** [comment]
FireSim quad-core rocket boots linux, powers off. I will bump FireSim tomorrow with updated AGFIs.

**Comment by davidbiancolin** [comment]
I've been placated. I too am always down for the net negative code change.

**Pull request merged**